packages/context/src/provider.ts

```diff
@@ -0,0 +1,103 @@
+// Copyright IBM Corp. 2013,2017. All Rights Reserved.
+// Node module: loopback
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Context} from './context';
+import {BoundValue, ValueOrPromise} from './binding';
+import {Constructor, instantiateClass} from './resolver';
+
+/**
+ * Provider of a value
+ */
+export interface Provider<T> {
+  value(ctx: Context): ValueOrPromise<T>;
+}
+
+/**
+ * BindingProvider provisions injecting a value into the binding asynchronously
+ * @example:
+ * ```ts
+ * ctx.bind('provider_key').toProvider(MyProvider);
+ * ```
+ */
+export abstract class BindingProvider<T> implements Provider<T> {
+  abstract source(): T;
+  abstract value(ctx: Context): ValueOrPromise<BoundValue>;
+}
+
+/**
+ * ValueProvider injects a value into the binding asynchronously
+ * @example:
+ * ```ts
+ * export class MyProvider extends ValueProvider {
+ *   constructor(@inject('value') param){}
+ *   value<string>(): string {
+ *     return computed_value;
+ *   }
+ * }
+ * ```
+ */
+export abstract class ValueProvider implements BindingProvider<ValueOrPromise<BoundValue>> {
+  abstract source(): ValueOrPromise<BoundValue>;
+  value(ctx: Context): ValueOrPromise<BoundValue> {
+    return this.source();
+  }
+}
+
+/**
+ * FunctionProvider injects a function into the binding asynchronously
+ * @example:
+ * ```ts
+ * export class MyProvider extends FunctionProvider {
+ *   constructor(@inject('value') param){}
+ *   value<() => string>(): () => string {
+ *     if (param)
+ *        return function1
+ *     else
+ *        return function2;
+ *   }
+ * }
+ * ```
+ */
+export abstract class FunctionProvider implements BindingProvider<() => ValueOrPromise<BoundValue>> {
+  abstract source(): () => ValueOrPromise<BoundValue>;
```

Comment:
Should it be `abstract source(): (ctx?: Context) => ValueOrPromise<BoundValue>;`?